kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+    /// encapsulated string.
+    ///
+    /// When a new leader is detected, this value will be updated with the leader's identity.
+    Following {
+        /// The ID of the current leader.
+        leader: String,
+        /// The last observed lease state.
+        lease: Lease,
+        /// The last time this lease state was observed.
+        last_updated: Instant,
+    },
+    /// A state indicating that the lease state is unknown, does not exist, or that the
+    /// corresponding leader elector task is starting or stopping.
+    Standby,
+}
+
+impl State {
+    /// Get a reference to the last known lease state.
+    fn get_lease(&self) -> Option<&Lease> {
+        match self {
+            Self::Leading { lease, .. } | Self::Following { lease, .. } => Some(lease),
+            Self::Standby => None,
+        }
+    }
+}
+
+/// Different states which a leader elector may be in.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum LeaderState {
+    /// This client instance is the leader.
+    Leading,
+    /// A state indicating that a different client is currently the leader.
+    Following,
+    /// A state indicating that the lease state is unknown, does not exist, or that the
+    /// corresponding leader elector task is starting or stopping.
+    Standby,
+}
+
+impl LeaderState {
+    /// Check if currently in `Leader` state.
+    #[allow(clippy::must_use_candidate)]
+    pub fn is_leader(&self) -> bool {
+        matches!(self, Self::Leading)
+    }
+}
+
+/// A handle to a leader elector task.
+pub struct LeaderElectorHandle {
+    /// Shutdown channel.
+    shutdown: oneshot::Sender<()>,
+    /// A watch signal over the observed leader state.
+    state: watch::Receiver<LeaderState>,
+    /// A join handle to the spawned leader elector task.
+    handle: JoinHandle<()>,
+}
+
+impl LeaderElectorHandle {
+    /// Get a handle to the state signal of this leader elector task.
+    ///
+    /// This signal receiver may be embedded in other parts of a program and used to govern actions
+    /// taken by the app in accordance with leader election state.
+    ///
+    /// Note that this leader elector task itself, as well as any other controller group members using
+    /// this same coordination implementation, will have an open watch on the lease object in the
+    /// Kubernetes API, and thus will see lease updates as they take place. This helps to ensure a
+    /// high degree of fencing, though it is not guaranteed.
+    #[allow(clippy::must_use_candidate)]
+    pub fn state(&self) -> watch::Receiver<LeaderState> {
+        self.state.clone()
+    }
+
+    /// Shutdown this leader elector task and return its underlying join handle.
+    #[allow(clippy::must_use_candidate)]
+    pub fn shutdown(self) -> impl Future<Output = Result<()>> {
+        let _res = self.shutdown.send(());
+        self.handle.map_err(|res| Error::TaskError(res.to_string()))
+    }
```

Comment:
This function should probably also vacate the Lease and null out the fields if we are the Leader.

Reply:
Do you mean that it should perform a final update of the lease in k8s as part of shutdown?

To be sure, with the currently implementation:
- The internal channel use for emitting LeaderState events will be closed, and consumers will get an error and will not be able to receive new events (the last state will still be present though).
- Other controllers will detect that the leader is no longer heartbeating and will then contend for leadership after the configured `lease_duration` elapses.

Reply:
Thoughts on this? One additional point to make is that it is actually nice to leave the state held in the lease as-is upon shutdown. This allows for a record of the last lease holder in k8s, which may be useful in some cases. 

Also, it is not necessary to clear the lease info. This system works well as-is for HA controller groups of 5 members or more (usually only a few are needed for HA setups), and other members of the HA group are still able to acquire leadership when they see that the old leader is no longer heartbeating the lease (falls outside of lease duration).